docs/design.md

```diff
@@ -0,0 +1,258 @@
+complex match pattern with different dependencies between each other.
+
+## Non-functional Requirements
+
+- The compliance checker should be packaged as a container image for easy
+  integration into CI/CD pipelines.
+
+## Design Choices
+
+### Data Model
+
+Based on the functional requirements, the [JSON schema][config-schema] for the rule
+configuration could look like this (in YAML for better readability):
+
+```yaml
+---
+$schema: https://json-schema.org/draft-07/schema
+additionalProperties: false
+definitions:
+  allOfRule:
+    additionalProperties: false
+    description: Asserts that all of the nested rules match.
+    properties:
+      allOf:
+        $ref: '#/definitions/rules'
+    required:
+      - allOf
+    type: object
+  anyOfRule:
+    additionalProperties: false
+    description: Asserts that at least one of the nested rules matches.
+    properties:
+      anyOf:
+        $ref: '#/definitions/rules'
+    required:
+      - anyOf
+    type: object
+  matchRule:
+    additionalProperties: false
+    description: A rule for matching trust anchors.
+    properties:
+      match:
+        description: |
+          Defines the pattern to match trust anchors against. Can be an exact
+          string or a regular expression.
+        type: string
+    required:
+      - match
+    type: object
+  notRule:
+    additionalProperties: false
+    description: Inverts the matching behaviour of a rule.
+    properties:
+      not:
+        $ref: '#/definitions/rule'
+    required:
+      - not
+    type: object
+  oneOfRule:
+    additionalProperties: false
+    description: Asserts that exactly one of the nested rules matches.
+    properties:
+      oneOf:
+        $ref: '#/definitions/rules'
+    type: object
+  rule:
+    description: Defines single matching rule.
+    oneOf:
+      - $ref: '#/definitions/allOfRule'
+      - $ref: '#/definitions/anyOfRule'
+      - $ref: '#/definitions/matchRule'
+      - $ref: '#/definitions/notRule'
+      - $ref: '#/definitions/oneOfRule'
+  rules:
+    description: A list of matching rules.
+    items:
+      $ref: '#/definitions/rule'
+    type: array
+description: Schema of the sops-compliance-checker configuration file
+properties:
+  rules:
+    $ref: '#/definitions/rules'
+title: sops-compliance-checker configuration
+type: object
+```
+
+This is highly influenced by the [JSON Schema Specification][jsonschema-spec]
+itself to allow for great flexibility in the rule composition.
+
+### Configuration Example
+
+Given the following requirements:
+
+- The SOPS file must contain all necessary approved trust anchors:
+  - AGE recovery key
+    `age1u79ltfzz5k79ex4mpl3r76p2532xex4mpl3z7vttctudr6gedn6ex4mpl3` used for
+    disaster recovery.
+  - AWS KMS keys of **at least one team** owning the component. Additional team
+    keys are allowed, as long as they are part of the list of approved trust
+    anchors.
+  - AWS KMS deploy keys for **exactly one** target environment
```

Comment:
```suggestion
  - AWS KMS keys used for deployment via CI/CD for **exactly one** target environment
```

There are no "deploy keys" in KMS, I'd suggest making it clear that we give those KMS keys another purpose which requires a different check